Add TransportHttpServer + Express + Opine, test scenarios

sgwilym opened the pull request:
## What's the problem you solved?

The `TransportHttpServer` class is only useful for Opine / Express servers.

## What solution are you recommending?

I've added a `TransportHttpHandler` class which has a `handler` property, a function with a signature of ```(req: Request) => Promise<Response>```, both classes from the [Fetch standard](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).

This function checks whether the request matches a specific URL pattern (e.g. `/from/:otherDeviceId`), executes the appropriate logic (e.g. getting a connection with a certain device, handling an envelope), and returns the result as a response.

Then you can do stuff like this:

```ts
import { serve } from "https://deno.land/std@0.123.0/http/server.ts";

const httpHandler = new SyncerHttpHandler(myPeer, "/");

await serve(httpHandler.handler);
```

And go on to compose the handler to intercept requests for a certain page, stuff like that.

I've also been thinking about how the handler could be composed into Opine / Express servers, but weirdly enough there are seemingly no utilities for converting express Requests to standard Requests, or standard Responses to express responses. If we had those, then you could do this:

```ts
app.on('*', async (req, res) => {
  const standardReq = toFetchReq(req);

  const standardRes = await httpHandler.handler(standardReq);

  applyStandardRes(res, standardRes);
});
```

Something like that.

Ultimately I think it's less work for us if we don't include an express / opine server with this library (and by extension, Earthstar...), but without a good way to convert standard fetch requests and results between frameworks that's a non-starter.

sgwilym committed:
Add a few little exports

sgwilym committed:
Add a TransportHTTPHandler

sgwilym committed:
Add TransportHttpHandler details to README

sgwilym commented:
(I'm working on polyfilling URLPattern — requires a PR of mine to be accepted: https://github.com/kenchris/urlpattern-polyfill/pulls)

cinnamon-bun commented:
Looks like this idea works well for Deno servers but not Node servers?  In any case I'd support merging it if we can get the node tests passing.  We might have to exclude it from the Node build if it won't work there anyway.

cinnamon-bun commented:
I'd also recommend renaming:

`TransportHttpHandler` --> `TransportHttpServerHandler`

I think each transport name should be clear about if it's a network server or network client (separate from whatever role it plays in the Earthstar protocol conversation)

sgwilym commented:
@cinnamon-bun Without readily available tooling for converting standard `Request` and `Response` objects to Express / Node equivalents, it's doable but annoying. I'm kind of surprised there isn't anything out there yet.

sgwilym committed:
TransportHttpHandler -> TransportHttpServerHandler

sgwilym committed:
Add URLPattern polyfill

sgwilym commented:
Renamed the class and now polyfilled for Node.

sgwilym committed:
Add connection tests for handler + express handler

sgwilym committed:
Temporary async bandaid

sgwilym committed:
What a day.

sgwilym committed:
Just import Deno for testing…

sgwilym commented:
Okay, here's my work today:

- Added `TransportHttpServerOpine` and `TransportHttpServerExpress`.
- Added different test scenarios for different platforms
  - Added a `ITransportScenario` class which describes how to set up and tear down a transport test
  - Added different transport scenarios for all the different transports
- Fiddled around with the NPM build and how things are required in order to make sure the NPM build doesn't get Opine / Deno stuff in it.
- I was dumbfounded for a few hours not realising that express was reusing the same server between tests 🙃🙃🙃

Node v16 tests work for me but Github is doing something weird with it. Node v14 tests has that weird eventtarget error which I thought we'd solved.

sgwilym committed:
Hm…

sgwilym committed:
Update dnt

sgwilym committed:
Cater to Node 14’s lack of fetch API

sgwilym committed:
Refactor HttpClient for rescheduling

sgwilym committed:
Catch requests being aborted

sgwilym commented:
I've refactored the HttpClient so that when it sends an envelope it reschedules the next pull of envelopes so that it happens much sooner.

This fixes an initial lull with the HttpClient where it would initialise, find no envelopes from the server (as they'd never spoken before), and then wait 1000ms. This was also making tests involving Http transports quite slow.

sgwilym committed:
Oh, don't forget this bit!

sgwilym committed:
Fix some leaks

sgwilym committed:
Use timers that adhere to spec

cinnamon-bun commented on README.md:
```diff
@@ -214,6 +216,8 @@ Server --> client: The client does GET requests to poll for batches of messages
 
 The server class takes an existing Express server app in its constructor, and adds a route handler to it. The client uses `fetch`.
 
+The handler class has a `handler` property with a function which takes a standard `Request` and returns a `Response.` This can be plugged as-is with many server frameworks, or you can compose the handler into something like an Express / Opine handler.
```
maybe clarify "as used by Fetch and by Deno's HTTP server"

cinnamon-bun commented on README.md:
```diff
@@ -206,6 +206,8 @@ If you try to do anything with a Connection or Transport that is CLOSED:
 
 `TransportHttpServer`
 
+`TransportHttpHandler`
```
This has a different name now

cinnamon-bun commented on src/test/connection.test.ts:
```diff
@@ -1,154 +1,149 @@
+    await t.step({
+        name: `notify + ${scenario.name}`,
+        // There may be async ops on the client and server
+        sanitizeOps: false,
```
Hopefully we can resolve these async ops, let's not forget about them, but ok to merge for now

sgwilym replied:
Because these steps happen over the lifetime of a server, the only way to make them not leak is to setup and close the client and server each and every test step. I think it's better to just set them up once, disable the leaking ops and comment specifically. Any leaking stuff will still be caught by the top-level test.

cinnamon-bun commented on src/test/connection.test.ts:
```diff
@@ -1,154 +1,149 @@
+            // because notify is not supposed to return anything.
+            // This will show as a console.warn but that's expected.
+
+            // error in the method call (no such method
```
my own typo!  missing end paren

cinnamon-bun commented on src/test/scenarios.node.ts:
```diff
@@ -0,0 +1,63 @@
+class TransportHttpExpressScenario<BagType extends FnsBag> implements ITransportScenario<BagType> {
+    name = 'TransportHttpClient + TransportHttpServerOpine';
+    connAtoB: IConnection<BagType>;
+    connBtoA: IConnection<BagType>;
```
could this be `IConnection<BagType> | null` so we don't have to do the convoluted typescript dance on line 31?

sgwilym replied:
This will mean that everywhere else will need to do null-checking for connBtoA. Acceptable?

cinnamon-bun commented on src/transport-http-client.ts:
```diff
@@ -58,28 +63,50 @@ export class TransportHttpClient<BagType extends FnsBag> implements ITransport<B
                     } else {
                         log('send... success.');
+
+                        // cancel current pull timers for this connection
```
Cancelling HTTP requests can be dangerous here - if we cancel a connection while it's transferring a large Envelope, I think that Envelope will be lost in the void since the server will believe it's sent.

sgwilym replied:
Fortunately this is not what happens! Rescheduling does not cancel the previous fetch, it just makes it so that the next poll comes sooner. The only time a fetch request is actually cancelled is when the connection is closed.

cinnamon-bun commented on src/transport-http-server.ts:
```diff
@@ -52,46 +49,77 @@ export class TransportHttpServer<BagType extends FnsBag> implements ITransport<B
 
             const envsToSend = this._outgoingBuffer.get(otherDeviceId) ?? [];
             log(`GET: sending ${envsToSend.length} envelopes`);
             this._outgoingBuffer.set(otherDeviceId, []);
```
This is a danger zone where the server can think it sent some envelopes and clear the buffer, but maybe the HTTP request gets interrupted and the envelopes are lost.  But I don't think the server has any way of knowing if the HTTP request was successfully completely sent back to the client...

cinnamon-bun replied:
I don't think it's catastrophic, the other end will eventually time out and throw a timeout error if any envelopes are lost.  The sync algorithm will need to handle that anyway.

sgwilym replied:
This is mitigated a bit by the HttpClient never actually cancelling a request itself.

cinnamon-bun commented on src/transport-http-client.ts:
```diff
@@ -100,46 +127,104 @@ export class TransportHttpClient<BagType extends FnsBag> implements ITransport<B
                     log(`got ${envs.length} envelopes`);
                     for (const env of envs) {
-                        if (this.isClosed) return;
+                        if (this.isClosed) return () => {};
```
why does this return an empty function, but line 150 just does a `return`?

sgwilym replied:
Leftover from a previous experiment. Changed back to `void`!

cinnamon-bun commented on src/transport-http-server.ts:
```diff
@@ -7,17 +7,13 @@ import { Connection } from './connection.ts';
 
-import type { Opine } from '../deps.ts';
-
-const TIMEOUT = 1000; // TODO: make this configurable
```
where is the general timeout defined now?

1 second is too short for production, but useful for testing

sgwilym replied:
I think the `TIMEOUT` exists only on the client, it's not needed in the server unless I'm mistaken.

cinnamon-bun commented on src/transport-http-server.ts:
```diff
@@ -102,18 +130,23 @@ export class TransportHttpServer<BagType extends FnsBag> implements ITransport<B
             deviceId: this.deviceId,
             methods: this.methods,
-            sendEnvelope: async (conn, env) => {
+            sendEnvelope: (_conn, env) => {
```
FYI the `conn` argument only exists here because in some other cases I needed to refer to the connection object from within `sendEnvelope`... while running the Connection constructor, so there was no easy way to get a handle on the connection object except by having it be passed into the function here.  Bit of a hack.

cinnamon-bun reviewed (approved):
Thank you so much for figuring out all the `dnt` and `npm` stuff 💜 

This looks good to me, feel free to merge at your leisure and address my comments afterwards if you want.

sgwilym committed:
Work in some feedback
